src/main/java/com/iota/iri/network/pipeline/BroadcastStage.java

```diff
@@ -0,0 +1,59 @@
+package com.iota.iri.network.pipeline;
+
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.network.NeighborRouter;
+import com.iota.iri.network.neighbor.Neighbor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * The {@link BroadcastStage} takes care of broadcasting newly received transactions to all neighbors except the
+ * neighbor from which the transaction originated from.
+ */
+public class BroadcastStage {
```

Comment:
I have noticed that you have many *Stage classes with the `process` method.
This is a good sign to add an interface (or abstract class, you choose).

Also if you declare `interface Stage<T>` and  `T process(T ctx)` then you can declare `BroadcastStage<BroadcastingContext>` and get rid of the cast.

Reply:
As discussed in DM, we can't actually write what we want to express with Generics in Java, because the stages return different types of payloads given the logic within the stage. Thereby, one always comes to casting as there's no other way to do it.